bootstrap: Don't trigger an unnecessary LLVM build from check builds

Zalathar opened the pull request:
Coming back to r-l/r development after a few weeks away, I found a major regression in my dev workflows: running `x check compiler` (either manually or via rust-analyzer) would have the side-effect of building LLVM, even though that shouldn't be necessary.

For my main build directory this would be a minor annoyance, but for my separate rust-analyzer build directory it's a huge problem because it causes a completely separate build of LLVM, which takes a long time and should be completely unnecessary.

---

After some investigation, I tracked down the problem to the `can_skip_build` check in this code:

https://github.com/rust-lang/rust/blob/3014e79f9c8d5510ea7b3a3b70d171d0948b1e96/src/bootstrap/src/core/build_steps/compile.rs#L1382-L1396

Historically, this would skip the LLVM build for stage 0 check builds. But after the recent stage 0 std redesign and some associated check stage renumbering (e.g. rust-lang/rust#143048), the condition `builder.top_stage == build_stage` is now false, because `top_stage` is 1 (due to the renumbering) but `build_stage` is 0 (because a “stage 1” non-library check build still uses the stage 0 compiler).

---

Because this is a critical contributor roadblock for me, I have tried to fix this in a relatively narrow way. It's possible that all of this surrounding logic could be greatly simplified (especially in light of the stage redesign changes), but I didn't want this fix to be held back by scope creep.

---

(Zulip thread: https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Bootstrap.20incorrectly.20building.20LLVM.20for.20check.20builds/near/528991035)

Zalathar committed:
Don't trigger an LLVM build from check builds using the stage 0 compiler

Zalathar commented:
r? bootstrap

Zalathar commented:
Hmm, in addition to those snapshot failures, I'm also seeing LLVM builds in subsequent manual testing. Mysterious.

@rustbot author

Zalathar commented:
Updated snapshot tests, several of which included the inappropriate LLVM build [(diff)](https://github.com/rust-lang/rust/compare/444b227a6afe701a9f04c44eeb51d15c17d13a2e..2aafc40e198c905ec80a9c29a33a39088e2fc0e3).

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16315853791/job/46081553520) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46081553520) [(plain)](https://github.com/rust-lang/rust/commit/857b6a2c8990c1e03f0e83fd5fe6d6e183e4c2d2/checks/46081553520/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Zalathar commented:
Having to deal with PR CI failures for `i686-pc-windows-gnu` feels like I am being pranked. Can't wait to see that target further demoted to oblivion.

Zalathar committed:
Clean up an unused compiler-stage parameter

Kobzol commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -630,6 +630,46 @@ fn any_debug() {
 }
 
+#[test]
+fn test_check_compiler_doesnt_build_llvm() {
```
I don't think we need a special test for this, because the existing snapshot tests already cover this (they check both `check library` and `check compiler` on multiple stages). Btw, I want to remove the `cache.contains::<step>` "API" and implement as much stuff as possible using snapshot tests.

Zalathar replied:
I wrote these tests before I knew about the snapshot suite, so I’m fine with removing them.

Kobzol reviewed (commented):
Thanks for taking a look! I was also looking at this recently, and was wondering whether we could just skip it completely for check builds. I thought that there might be some problem with `rustc_llvm` trying to use the built LLVM, but it looks like it's.. fine?

Zalathar committed:
Split some multi-snapshot tests to make blessing easier

When a snapshot test fails, it only emits a `.pending-snap` file for the first
snapshot assertion that actually failed, because subsequent assertions aren't
executed. That makes it cumbersome to re-bless tests that contain multiple
snapshot assertions.

Zalathar commented:
Removed the non-snapshot tests, and added some more comments to explain my understanding of why the LLVM conditional is tricky [(diff)](https://github.com/rust-lang/rust/compare/bf75472456c78cae47f190f33cef1e3645090ded..a028ca5f9bd771e825e5c8c3c55f938aea89985e).

Zalathar commented:
> Thanks for taking a look! I was also looking at this recently, and was wondering whether we could just skip it completely for check builds. I thought that there might be some problem with `rustc_llvm` trying to use the built LLVM, but it looks like it's.. fine?

My understanding is that it's better for check and non-check builds to use the same environment, if possible, to avoid unnecessary rebuilds due to cache-busting.

I haven't checked whether that's actually how it works in practice, but it seems plausible.

Kobzol commented:
Oh, right, cache busting. I don't actually understand how does the cache get busted here, particularly because check and build don't share caches at all (like, in general, in Cargo, not just in bootstrap)? If it's busting between doing `check`, `build`, `check`, then LLVM should only be built in an expensive way once, so the followup builds shouldn't skip it. So I think/hope that should be fine. I did a few check/build/check combos and saw no rebuilds, at least on stage 1.

What I *think* that https://github.com/rust-lang/rust/pull/110122 was about is that no LLVM was being built for what we call a "stage 2" check build now. I tried to do that with `x check --stage 2` and an empty build dir, and it seems like it works fine now.

Zalathar commented:
Oh whoops, I forgot to mark it ready.

@rustbot ready

Kobzol commented:
@bors r+

bors commented:
:pushpin: Commit a028ca5f9bd771e825e5c8c3c55f938aea89985e has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol a028ca5f9bd771e825e5c8c3c55f938aea89985e -->
<!-- homu: {"type":"Approved","sha":"a028ca5f9bd771e825e5c8c3c55f938aea89985e","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
